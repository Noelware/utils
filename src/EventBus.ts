/*
 * ğŸŒ¸ @noelware/utils: Noelware's utilities package to not repeat code in our TypeScript projects.
 * Copyright (c) 2021-2022 Noelware <team@noelware.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/** Represents a generic listener function. */
export type Listener = (...args: unknown[]) => void;

/** Extracts the arguments of a specific {@link Listener}. */
export type ExtractListenerArguments<L> = L extends Listener ? Parameters<L> : any[];

/** Represents a generic definition of the type-safe events in a {@link EventBus}. */
export type GenericEventBusMap = Record<string, Listener>;

export interface EventEmitterLike {
  addListener(event: string, listener: (...args: any[]) => void): any;
  emit(event: string, ...args: unknown[]): any;
  once(event: string, listener: (...args: any[]) => void): any;
  on(event: string, listener: (...args: any[]) => void): any;
}

/**
 * Checks if {@link emitter} is a event emitter or not.
 * @param emitter The unknown property.
 * @returns Asserts that {@link emitter} is {@link EventEmitterLike}.
 */
export function isEventEmitterLike<T extends EventEmitterLike>(emitter: unknown): emitter is T {
  const conditions = [
    typeof emitter === 'object' && !Array.isArray(emitter) && emitter !== null,
    typeof (emitter as EventEmitterLike).addListener === 'function',
    typeof (emitter as EventEmitterLike).emit === 'function',
    typeof (emitter as EventEmitterLike).once === 'function',
    typeof (emitter as EventEmitterLike).on === 'function'
  ];

  return conditions.every((t) => t === true);
}

/**
 * Represents an extended event emitter.
 */
export class EventBus<Map extends {} = GenericEventBusMap> {
  ['constructor']!: typeof EventBus;
  #maxListenerSize = 250;
  #listeners: Record<keyof Map, Listener[]> = {} as any;

  /**
   * Emits a event from the listener callstack
   * @param event The event to be emitted.
   * @param args The arguments for the event.
   */
  emit<E extends keyof Map>(event: E, ...args: ExtractListenerArguments<Map[E]>) {
    if (!this.#listeners.hasOwnProperty(event)) return;

    const listeners = this.#listeners[event];
    if (!listeners.length) return;
    for (const listener of listeners) {
      listener(...args);
    }
  }

  /**
   * Sets the max listeners to {@link count}. Use `-1` if you wish to not have
   * a max listener count, this is dangerous.
   * @param count The amount of listeners this {@link EventBus} can hold.
   * @returns This {@link EventBus} instance to chain methods
   */
  setMaxListeners(count: number) {
    this.#maxListenerSize = count;
    return this;
  }

  /**
   * Adds a new listener to the event's callstack.
   * @param event The event to add
   * @param listener The listener callback when it is emitted.
   * @return This {@link EventBus} instance to chain methods
   */
  on<E extends keyof Map>(event: E, listener: Map[E]) {
    const listeners = this.#listeners[event] ?? [];
    if (this.#maxListenerSize !== -1 && listeners.length > this.#maxListenerSize)
      throw new RangeError(`Reached the maximum amount of listeners to append on event [${String(event)}]`);

    listeners.push(listener as any);
    this.#listeners[event] = listeners;

    return this;
  }

  /**
   * Adds a new listener to the event's callstack that will be emitted once and removed
   * from the callstack.
   * @param event The event to register
   * @param listener The listener callback when it is emitted.
   * @return This {@link EventBus} instance to chain methods
   */
  once<E extends keyof Map>(event: E, listener: Map[E]) {
    const listenerToEmit = (...args: unknown[]) => {
      (listener as any)(...args);
      this.removeListener(event, listenerToEmit as unknown as Map[E]);
    };

    return this.on(event, listenerToEmit as unknown as Map[E]);
  }

  /**
   * Removes a callback listener from the event's callstack.
   * @param event The event
   * @param listener The listener callback that was registered.
   */
  removeListener<E extends keyof Map>(event: E, listener: Map[E]) {
    if (!this.#listeners.hasOwnProperty(event)) return false;

    const listeners = this.#listeners[event as string];
    if (!listeners.length) return;

    const index = listeners.indexOf(listener as any);
    if (index !== -1) listeners.splice(index, 1);

    this.#listeners[event] = listeners;
  }

  /**
   * Returns how many event callbacks have been registered.
   * @param event The event to check
   */
  size<E extends keyof Map>(event: E): number;

  /**
   * Returns how many events have been registered.
   */
  size(): number;
  size(event?: string) {
    if (event !== undefined) {
      const listeners = this.#listeners[event] ?? [];
      return listeners.length;
    }

    return Object.keys(this.#listeners).length;
  }

  /**
   * Removes all the listeners from this {@link EventBus}
   * @return This {@link EventBus} instance to chain methods
   */
  removeAllListeners() {
    this.#listeners = {} as any;
    return this;
  }

  /** @inheritdoc EventBus.on */
  addListener<E extends keyof Map>(event: E, listener: Map[E]) {
    return this.on(event, listener);
  }
}
