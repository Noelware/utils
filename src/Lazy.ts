/*
 * ðŸŒ¸ @noelware/utils: Noelware's utilities package to not repeat code in our TypeScript projects.
 * Copyright (c) 2021-2024 Noelware, LLC. <team@noelware.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * Creates a new {@link Lazy} object from a function.
 *
 * Since 2.5.1: lazily functions can include arguments as well
 *
 * @param func The lazy function.
 * @returns A {@link Lazy} object.
 */
export function lazy<T, Args extends any[] = any[]>(func: (...args: Args) => T): Lazy<T, Args> {
    return new Lazy(func);
}

/**
 * Represents a state object to calculate something in a function
 * lazily and return the same output each time.
 *
 * ## Example
 * ```ts
 * import { lazy } from '@noelware/utils';
 *
 * const myExpensiveFunc = lazy(() => 42);
 * console.log(myExpensiveFunc.get()); // => 42
 * console.log(myExpensiveFunc.get()); // => 42
 * ```
 */
export class Lazy<T, Args extends any[] = any[]> {
    #cached: T | undefined = undefined;
    #func: (...args: Args) => T;

    constructor(func: (...args: Args) => T) {
        this.#func = func;
    }

    get(...args: Args) {
        return this.#cached !== undefined ? this.#cached : (this.#cached = this.#func.call(this, ...args));
    }
}
