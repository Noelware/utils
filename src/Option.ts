/*
 * ðŸŒ¸ @noelware/utils: Noelware's utilities package to not repeat code in our TypeScript projects.
 * Copyright (c) 2021-2024 Noelware, LLC. <team@noelware.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const kVal = Symbol.for('$hoshi::option::value$');

/**
 * Represents a Rust-style [`Option`] interface. This mirrors some APIs that Rust's Option
 * type brings in. You use the {@link some `some()`} and {@link none `none()`} methods to create
 * an instance of this instead of calling `new Option(...)`.
 *
 * @since 2.5.0
 *
 * ## Example
 * ```ts
 * import { some, none } from '@noelware/utils/option';
 *
 * const meaningOfLife = some(42);
 * const nil = none();
 *
 * nil.inspect(() => {});
 * // the closure is never called
 *
 * meaningOfLife.inspect(val => console.log(`value is ${val}`));
 * // the closure is called and prints 'value is 42'
 * ```
 *
 * [`Option`]: https://doc.rust-lang.org/nightly/core/option/enum.Option.html
 */
export class Option<T> {
    [kVal]: T | undefined;

    /**
     * Returns a `boolean` to indicate that this {@link Option `Option`} resolves to something
     * meaningful.
     *
     * @since 2.5.0
     *
     * ## Example
     * ```ts
     * import { some } from '@noelware/utils/option';
     *
     * const variant = some(42);
     * variant.isSome();
     * // => true
     * ```
     */
    isSome() {
        return this[kVal] !== undefined;
    }

    /**
     * Returns a `boolean` to indicate if this {@link Option `Option`} is nothing.
     *
     * @since 2.5.0
     *
     * ## Example
     * ```ts
     * import { none } from '@noelware/utils/option';
     *
     * const variant = none();
     * variant.isNone();
     * // => true
     * ```
     */
    isNone() {
        return !this.isSome();
    }

    /**
     * Maps this `Option<T>` into a `Option<U>` with applying a function
     * to the contained Some variant, leaving the None variant untouched.
     *
     * @since 2.5.0
     *
     * ## Example
     * ```ts
     * import { some } from '@noelware/utils/option';
     *
     * const variant = some(42);
     * //    ^? (Option<number>)
     *
     * const myNewVariant = variant.map((value) => value * 2);
     * myNewVariant.inspect((value) => {
     *     console.log(`my new value is ${value}`);
     *     // will print 'my new value is 84'
     * });
     * ```
     *
     * @param fn Function to converse `T` ~> `U`.
     * @returns a new {@link Option `Option`} type of the contained `U` passed in.
     */
    map<U>(fn: (value: T) => U): Option<U> {
        if (this.isSome()) {
            return some(fn(this[kVal]!));
        }

        return this as unknown as Option<U>;
    }

    /**
     * Calls the provided `fn` closure to inspect the Some variant. If this is a None variant,
     * this will do nothing.
     *
     * @since 2.5.0
     *
     * ## Examples
     * ```ts
     * import { some } from '@noelware/utils/option';
     *
     * let x = some(4)
     *     .inspect((value) => console.log(`original: ${value}`))
     *     .map((value) => value * 3)
     *     .unwrap();
     * ```
     *
     * @param fn Closure to call to inspect the `value`.
     * @returns this instance to chain methods
     */
    inspect(fn: (value: T) => void) {
        if (this.isSome()) {
            fn(this[kVal]!);
        }

        return this;
    }

    /**
     * Unwrap a Some variant's value or throws an exception if this Option
     * was a none variant.
     *
     * @since 2.5.0
     */
    unwrap(): T {
        if (this.isNone()) {
            throw new Error('called Option#unwrap() on a `None` value');
        }

        return this[kVal]!;
    }

    /**
     * Unwrap a Some variant's value or throws an exception (with a custom message) if this Option
     * was a none variant.
     *
     * @param message The message to throw
     * @since 2.5.0
     */
    expect(message: string): T {
        if (this.isNone()) {
            throw new Error(message);
        }

        return this[kVal]!;
    }
}

/**
 * Resolves a value that can represent something.
 * @param value The value to contain.
 * @since 2.5.0
 */
export function some<T>(value: T) {
    const option = new Option<T>();
    option[kVal] = value;

    return option;
}

/**
 * Creates a option that resolves nothing.
 * @since 2.5.0
 */
export function none<T>(): Option<T> {
    return new Option();
}
