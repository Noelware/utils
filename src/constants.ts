/*
 * ðŸŒ¸ @noelware/utils: Noelware's utilities package to not repeat code in our TypeScript projects.
 * Copyright (c) 2021-2023 Noelware <team@noelware.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

declare var window: any;
declare var document: any;

/** Months represented as 0-indexed values to the month name */
export const Months: { [month: number]: string } = {
    0: 'January',
    1: 'Feburary',
    2: 'March',
    3: 'April',
    4: 'May',
    5: 'June',
    6: 'July',
    7: 'August',
    8: 'September',
    9: 'October',
    10: 'November',
    11: 'December'
};

/** The days of a week */
export const DaysInWeek: { [day: number]: string } = {
    0: 'Sunday',
    1: 'Monday',
    2: 'Tuesday',
    3: 'Wednesday',
    4: 'Thursday',
    5: 'Friday',
    6: 'Saturday'
};

/**
 * Returns if we are in the browser or not
 */
export const isBrowser = (() => {
    const attributes = [typeof window, typeof document];
    return attributes.every((i) => i !== 'undefined');
})();

/** Returns if we are in a Node.js environment or not. */
export const isNode = typeof process !== 'undefined' && !isBrowser;
