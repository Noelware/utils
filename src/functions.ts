/*
 * ðŸŒ¸ @noelware/utils: Noelware's utilities package to not repeat code in our TypeScript projects.
 * Copyright (c) 2021-2022 Noelware <team@noelware.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { OmitUndefinedOrNull, ReaddirOptions } from './types';

declare var window: any;

let cachedModules = new Map<string, object>();

/**
 * Returns if we are in the browser or not
 */
export const isBrowser = typeof process === 'undefined' && typeof window === 'object';

/** Returns if we are in a Node.js environment or not. */
export const isNode = typeof process !== 'undefined' && typeof window === 'undefined';

/**
 * Calculates the distance of `process.hrtime` and
 * returns the milliseconds from the duration
 *
 * @param start The start array of [seconds, nanoseconds]
 * @returns The time in milliseconds
 */
export function calculateHRTime(start: [seconds: number, nanoseconds: number]) {
  const difference = process.hrtime(start);
  return (difference[0] * 1e9 + difference[1]) / 1e6;
}

/**
 * Sleeps on the current promise for a specific amount of time.
 * @param duration The duration to sleep on.
 */
export function sleep(duration: number) {
  return new Promise<unknown>((resolve) => setTimeout(resolve, duration));
}

/**
 * Returns if the key is available in the object.
 * @param obj The object to check
 * @param key The key to check if it exists.
 */
export function hasOwnProperty<T extends {}, K extends keyof T = keyof T>(obj: T, key: K) {
  return Object.prototype.hasOwnProperty.call(obj, key);
}

/**
 * Checks if {@link obj} is a object type. This doesn't return true for arrays or `null`.
 * @param obj The object to check.
 */
export function isObject<T extends {} = {}>(obj: unknown): obj is T {
  if (Array.isArray(obj)) return false;
  if (obj === null) return false;
  if (typeof obj !== 'object') return false;

  return true;
}

/**
 * Returns the property if it's found in a {@link object obj}, otherwise it defaults to the {@link default defaultValue}.
 * @param obj The object to check
 * @param key The key
 * @param defaultValue The default value, if it wasn't found
 */
export function property<T extends {}, K extends keyof T = keyof T>(obj: T, key: K, defaultValue: T[K]): T[K] {
  if (!isObject(obj)) return defaultValue;
  if (!hasOwnProperty(obj, key)) return defaultValue;
  return obj[key];
}

/**
 * Helper function to pluralize a number.
 * @param str The string to use when pluralizing the number.
 * @param value The actual number value.
 */
export function pluralize(str: string, value: number) {
  if (value === 0) return `${value} ${str}s`;
  return value >= 1.5 ? `${value} ${str}s` : `${value} ${str}`;
}

/**
 * Humanizes a duration via {@link ms} and returns the humanized version.
 * @param ms The duration to use
 * @param long If the output should be `$num $duration` (i.e, `1 year`) or not.
 */
export function humanize(ms: number, long: boolean = false) {
  const years = Math.floor(ms / 31104000000);
  const months = Math.floor((ms / 2592000000) % 12);
  const weeks = Math.floor((ms / 604800000) % 7);
  const days = Math.floor((ms / 86400000) % 30);
  const hours = Math.floor((ms / 3600000) % 24);
  const minutes = Math.floor((ms / 60000) % 60);
  const seconds = Math.floor((ms / 1000) % 60);

  const strings: string[] = [];
  if (years > 0) strings.push(long ? pluralize('year', years) : `${years}y`);
  if (months > 0) strings.push(long ? pluralize('month', months) : `${months}mo`);
  if (weeks > 0) strings.push(long ? pluralize('week', weeks) : `${weeks}w`);
  if (days > 0) strings.push(long ? pluralize('day', days) : `${days}d`);
  if (hours > 0) strings.push(long ? pluralize('hour', hours) : `${hours}h`);
  if (minutes > 0) strings.push(long ? pluralize('minute', minutes) : `${minutes}m`);
  if (seconds > 0) strings.push(long ? pluralize('second', seconds) : `${seconds}s`);

  return strings.filter(Boolean).join(long ? ', ' : '');
}

/**
 * Manipulates a string's text with the first letter being uppercase and the rest
 * being untouched.
 *
 * @param text The text to use
 * @param delim The delimiter to split, defaults to a space.
 */
export function titleCase(text: string, delim: string = ' ') {
  return text
    .split(delim)
    .map((t) => `${t.charAt(0).toUpperCase()}${t.slice(1)}`)
    .join(' ');
}

/**
 * Checks if the item in the data can be excluded via a predicate function.
 * @param data The data array to use
 * @param predicate The predicate function to check if it can be excluded.
 */
export function shouldExclude<T>(data: T[], predicate: (item: T) => boolean) {
  if (!data.length) return false;
  return data.some(predicate);
}

/**
 * Lazy evaluates a module via `require` and caches it, or returns the cached result
 * if the module was already lazily evaluated.
 *
 * @param mod The module to evaluate
 * @returns The module as {@link T}.
 */
export function tryRequire<T>(mod: string): T {
  try {
    if (cachedModules.has(mod)) return cachedModules.get(mod) as unknown as T;
    const m = require(mod);

    cachedModules.set(mod, m);
    return m;
  } catch (e) {
    assertIsError(e);

    throw new Error(`Module "${mod}" was not found. Did you install it?`, { cause: e });
  }
}

/**
 * Reads a directory's contents and returns the contents and the subdirectories' children.
 * @param path The path to look for, this can be relative or absolute.
 * @param options The {@link ReaddirOptions} options to use, if needed.
 * @returns A list of files if in a Node.js environment, otherwise an empty array if in a browser context.
 */
export function readdirSync(path: string, options: ReaddirOptions = {}) {
  // If it is the browser, this will be a empty array since you can't read directories.
  if (isBrowser) return [];

  const extensions = property(options, 'extensions', []) ?? [];
  const excluded = property(options, 'exclude', []) ?? [];

  let results: string[] = [];
  const fs = tryRequire<typeof import('fs')>('fs');
  const { join } = tryRequire<typeof import('path')>('path');
  const excludePredicate = (file: import('fs').Dirent) => (item: string | RegExp) => {
    if (typeof item === 'string') return item === file.name;
    return item.test(file.name);
  };

  const files = fs.readdirSync(path, { withFileTypes: true });
  for (let i = 0; i < files.length; i++) {
    const file = files[i];
    const fullPath = join(path, file.name);

    const stat = fs.statSync(fullPath);
    const isFile = stat.isFile() || stat.isFIFO();
    if (isFile) {
      if (shouldExclude(excluded, excludePredicate(file))) continue;
      if (shouldExclude(extensions, excludePredicate(file))) continue;

      results.push(file.name);
    } else {
      if (shouldExclude(excluded, excludePredicate(file))) continue;
      if (shouldExclude(extensions, excludePredicate(file))) continue;

      results = results.concat(...readdirSync(path, options));
    }
  }

  return results;
}

export function readdir(path: string, options: ReaddirOptions = {}) {
  // If it is the browser, this will be a empty array since you can't read directories.
  if (isBrowser) return Promise.resolve([]);
  return new Promise<string[]>((resolve) => {
    const results = readdirSync(path, options);
    resolve(results);
  });
}

export function omitUndefinedOrNull<T extends object>(obj: T) {
  return Object.keys(obj).reduce<OmitUndefinedOrNull<T>>((acc, curr) => {
    if (obj[curr] === undefined || obj[curr] === null) return acc;

    acc[curr] = obj[curr];
    return acc;
  }, {} as OmitUndefinedOrNull<T>);
}

export function assertIsError<E extends Error = Error>(value: unknown): asserts value is E {
  if (!(value instanceof Error)) throw new Error(`Value was not \`Error\`.`);
}
