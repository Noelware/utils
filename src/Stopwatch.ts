/*
 * ðŸŒ¸ @noelware/utils: Noelware's utilities package to not repeat code in our TypeScript projects.
 * Copyright (c) 2021-2022 Noelware <team@noelware.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { tryRequire } from './functions';
import { isBrowser } from './constants';
import { lazy } from './Lazy';

const now = lazy(() =>
  isBrowser
    ? Date.now
    : (() => {
        const mark = tryRequire<typeof import('perf_hooks')>('perf_hooks');
        return mark.performance.now;
      })()
);

export class Stopwatch {
  #startTime?: number;
  #now: () => number = now.get();
  #endTime?: number;

  static createStarted() {
    const self = new Stopwatch();
    self.start();

    return self;
  }

  static measure(func: () => void) {
    const self = Stopwatch.createStarted();
    func();

    return self.stop();
  }

  static async measureAsync(func: () => Promise<void>) {
    const self = Stopwatch.createStarted();
    await func();

    return self.stop();
  }

  start() {
    if (this.#startTime !== undefined) return;
    this.#startTime = this.#now();
  }

  stop() {
    if (this.#startTime === undefined) return;
    this.#endTime = this.#now();

    if (this.#endTime > 1000) return `${this.#endTime.toFixed(1)}s`;
    if (this.#endTime > 1) return `${this.#endTime.toFixed(1)}ms`;
    return `${this.#endTime.toFixed(1)}Âµs`;
  }
}
