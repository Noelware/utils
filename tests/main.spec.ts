/*
 * ðŸŒ¸ @noelware/utils: Noelware's utilities package to not repeat code in our TypeScript projects.
 * Copyright (c) 2021-2023 Noelware <team@noelware.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { assert, beforeEach, expect, test, describe } from 'vitest';
import * as utils from '../src';

test('constants', () => {
    expect(utils.isNode).toBe(true);
    expect(utils.isBrowser).toBe(false);

    assert.deepEqual(
        {
            0: 'Sunday',
            1: 'Monday',
            2: 'Tuesday',
            3: 'Wednesday',
            4: 'Thursday',
            5: 'Friday',
            6: 'Saturday'
        },
        utils.DaysInWeek
    );

    assert.deepEqual(
        {
            0: 'January',
            1: 'Feburary',
            2: 'March',
            3: 'April',
            4: 'May',
            5: 'June',
            6: 'July',
            7: 'August',
            8: 'September',
            9: 'October',
            10: 'November',
            11: 'December'
        },
        utils.Months
    );
});

describe('EventBus', () => {
    const eventBus = new utils.EventBus();
    beforeEach(() => {
        eventBus.removeAllListeners();
    });

    test('can emit', () => {
        let emitted = false;
        eventBus.on('owo', () => {
            emitted = true;
        });

        eventBus.emit('owo');
        expect(emitted).toBeTruthy();
    });

    test('has 100 event listeners on the `owo` event', () => {
        for (let i = 0; i < 100; i++) {
            eventBus.on('owo', () => {
                // noop
            });
        }

        expect(eventBus.size()).toBe(1);
        expect(eventBus.size('owo')).toBe(100);
    });

    test('should be no listeners with `once` method', () => {
        let hasRemoved = false;
        eventBus.once('uwu', () => {
            hasRemoved = true;
        });

        expect(eventBus.size()).toBe(1);
        eventBus.emit('uwu');

        expect(hasRemoved).toBeTruthy();
        expect(eventBus.size()).toBe(1);
        expect(eventBus.size('uwu')).toBe(0);
    });

    test('should now have zero event listeners', () => {
        expect(eventBus.size()).toBe(0);
        expect(eventBus.size('owo')).toBe(0);
    });
});

test('Lazy', () => {
    const lazyValue = new utils.Lazy(() => 1);
    expect(lazyValue.get()).toBe(1);
    expect(lazyValue.get()).toBe(1);
});

describe('functions', () => {
    test('hasOwnProperty', () => {
        const obj: Record<string, any> = { uno: 1, dos: 2 };
        expect(utils.hasOwnProperty(obj, 'uno')).toBeTruthy();
        expect(utils.hasOwnProperty(obj, 'dos')).toBeTruthy();
        expect(utils.hasOwnProperty(obj, 'hssksidjslksdjd')).toBeFalsy();
    });

    test('isObject', () => {
        expect(utils.isObject({ owo: true })).toBeTruthy();
        expect(utils.isObject(null)).toBeFalsy();
        expect(utils.isObject(['owo'])).toBeFalsy();
        expect(utils.isObject('nmdskdsdsla;dlsdsks;aldksdl;ad')).toBeFalsy();
    });

    test('titleCase', () => {
        expect(utils.titleCase('everything is owo')).toBe('Everything Is Owo');
    });

    test('shouldExclude', () => {
        expect(utils.shouldExclude(['h', 'i', 'j'], (i) => i === 'h')).toBeTruthy();
        expect(utils.shouldExclude(['a', 'b', 'c'], (i) => i.length === 3)).toBeFalsy();
    });

    test('omitUndefinedOrNull', () => {
        const items = { a: undefined, b: null, c: 'a', d: 1, e: true };
        expect(utils.omitUndefinedOrNull(items)).toMatchInlineSnapshot(`
      {
        "c": "a",
        "d": 1,
        "e": true,
      }
    `);
    });

    test('assertIsError', () => {
        function isError(value: unknown) {
            try {
                utils.assertIsError(value);
                return true;
            } catch (e) {
                throw e;
            }
        }

        expect(() => isError('abcd')).toThrowError();
        expect(() => isError(new Error('beep boop'))).not.toThrowError();
    });
});

describe('Stopwatch', () => {
    const _ = new utils.Stopwatch();
    test('1 + 1 = 2?', () => expect(1 + 1).toBe(2));
});
