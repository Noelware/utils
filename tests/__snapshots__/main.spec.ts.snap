// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`functions omitUndefinedOrNull 1`] = `
{
  "c": "a",
  "d": 1,
  "e": true,
}
`;
